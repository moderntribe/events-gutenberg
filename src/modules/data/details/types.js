export const SET_DETAILS = 'SET_DETAILS';
export const SET_POST_TYPE = 'SET_POST_TYPE';
export const SET_LOADING = 'SET_LOADING';
