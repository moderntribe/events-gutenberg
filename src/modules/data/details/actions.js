/**
 * Internal dependencies
 */
import * as types from './types';

export const enableIsLoading = ( id ) => ( {
	type: types.SET_IS_LOADING,
	payload: {
		id,
		loading: true,
	},
} );

export const disableIsLoading = ( id ) => ( {
	type: types.SET_IS_LOADING,
	payload: {
		id,
		loading: false,
	},
} );

export const setDetails = ( id, details ) => ( {
	type: types.SET_DETAILS,
	payload: {
		id,
		details,
	},
} );

export const setPostType = ( id, postType ) => ( {
	type: types.SET_POST_TYPE,
	payload: {
		id,
		postType,
	},
} );
