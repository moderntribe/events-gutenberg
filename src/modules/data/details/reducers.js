/**
 * External dependencies
 */
import moment from 'moment/moment';

/**
 * Internal dependencies
 */
import {
	toDateTime,
	roundTime,
	isSameDay,
	setTimeInSeconds,
	totalSeconds,
	replaceDate,
	toMoment,
} from 'utils/moment';

import {
	HALF_HOUR_IN_SECONDS,
	DAY_IN_SECONDS,
} from 'utils/time';

import { isTruthy } from 'utils/string';

/**
 * Set the event to be a multi day event
 *
 * @param {object} prevState Previous state
 * @param {boolean} multiDay If is multiday or not.
 * @returns {object} new state object
 */
export function setMultiDay( prevState, multiDay ) {
	const { startDate, endDate } = prevState;

	if ( multiDay ) {
		return getMultiDayDates( prevState );
	}

	/**
	 * If we are no longer on a range and start and end are not on the same day move back the dates
	 * to the same day using the start date as the rule to determine the dates.
	 *
	 * after this point we can assure we are setting multiday to off
	 */
	const start = toMoment( startDate );
	const state = {
		multiDay,
		startDate,
		endDate,
	};

	const sameDay = isSameDay( start, toMoment( endDate ) );
	if ( ! sameDay ) {
		const lastBlockOfTheDay = DAY_IN_SECONDS - HALF_HOUR_IN_SECONDS;
		// Make sure we don't overflow if start is on the last block of the day which is 11:30
		// As if we just add 30 more minutes we are going to overflow again.
		if ( totalSeconds( start ) === lastBlockOfTheDay ) {
			state.startDate = toDateTime( start.subtract( HALF_HOUR_IN_SECONDS, 'seconds' ) );
		}
		state.endDate = toDateTime(
			toMoment( state.startDate ).add( HALF_HOUR_IN_SECONDS, 'seconds' ),
		);
	}

	state.allDay = isAllDay( toMoment( state.startDate ), toMoment( state.endDate ) );

	return {
		...prevState,
		...state,
	};
}

/**
 * Set the event to be an All Day event
 *
 * @param {object} prevState Previous state
 * @param {boolean} allDay If the event should be all day
 * @returns {object} new state object
 */
export function setAllDay( prevState, allDay ) {
	if ( allDay ) {
		return getAllDayDates( prevState );
	}

	const { startDate } = prevState;
	return setStartDate( prevState, startDate );
}

/**
 * Change the time of the start date based on an amount of seconds
 *
 * @param {object} prevState Previous state
 * @param {number} seconds Total of seconds to set the start day
 * @returns {object} new state object
 */
export function setStartTime( prevState, seconds ) {
	const { allDay } = prevState;
	if ( allDay ) {
		return getAllDayDates( prevState );
	}

	const { startDate } = prevState;
	const start = setTimeInSeconds( toMoment( startDate ), seconds );
	return setStartDate( prevState, toDateTime( start ) );
}

/**
 * Change the time of the end date based on an amount of seconds
 *
 * @param {object} prevState Previous state
 * @param {number} seconds Total of seconds to be set the end date
 * @returns {object} new state object
 */
export function setEndTime( prevState, seconds ) {
	const { allDay } = prevState;
	if ( allDay ) {
		return getAllDayDates( prevState );
	}

	const { startDate } = prevState;
	const end = setTimeInSeconds( toMoment( startDate ), seconds );
	return setEndDate( prevState, toDateTime( end ) );
}

export function setInitialState( prevState, values ) {
	const state = setInitialDates( values, values.startDate, values.endDate );

	return {
		...prevState,
		...state,
	};
}

export function setInitialDates( prevState = {}, startDate = '', endDate = '' ) {
	let start = toMoment( startDate );
	let end = toMoment( endDate );

	if ( ! start.isValid() ) {
		start = roundTime( moment() );
	}

	if ( ! end.isValid() ) {
		end = roundTime( moment() );
	}

	const state = {
		...prevState,
		endDate: toDateTime( end ),
	};
	return state.allDay ? state : setStartDate( state, toDateTime( start ) );
}

export function setOnlyDateForStart( prevState, date ) {
	if ( ! date ) {
		return prevState;
	}

	const { startDate } = prevState;
	const start = replaceDate(
		toMoment( startDate ),
		roundTime( toMoment( date ) ),
	);

	return setStartDate( prevState, toDateTime( start ) );
}

export function setOnlyDateForEnd( prevState, date ) {
	if ( ! date ) {
		return prevState;
	}

	const { endDate } = prevState;
	const end = replaceDate(
		toMoment( endDate ),
		roundTime( toMoment( date ) ),
	);

	return setEndDate( prevState, toDateTime( end ) );
}

export function setStartDate( prevState, date ) {
	const { endDate } = prevState;
	const state = {
		startDate: roundTime( toMoment( date ) ),
		endDate: roundTime( toMoment( endDate ) ),
	};

	const start = roundTime( toMoment( date ) );
	if ( state.endDate.isSameOrBefore( start ) ) {
		state.endDate = start.add( HALF_HOUR_IN_SECONDS, 'seconds' );
	}

	return {
		...prevState,
		...adjustEventPropertiesFromMoment( state ),
	};
}

function isAllDay( startDate, endDate ) {
	const sameDay = isSameDay( toMoment( startDate ), toMoment( endDate ) );

	if ( ! sameDay ) {
		return false;
	}

	const isStartOfTheDay = totalSeconds( startDate ) === 0;
	const isEndOfTheDay = totalSeconds( endDate ) === ( DAY_IN_SECONDS - 1 );
	return isStartOfTheDay && isEndOfTheDay;
}

function setEndDate( prevState, date ) {
	const { startDate } = prevState;
	const state = {
		startDate: roundTime( toMoment( startDate ) ),
		endDate: roundTime( toMoment( date ) ),
	};

	if ( state.endDate.isBefore( state.startDate ) ) {
		// Swap dates if end date is before start date
		const [ end, start ] = [ state.startDate, state.endDate ];
		state.startDate = start;
		state.endDate = end;
	}

	if ( state.endDate.isSame( state.startDate ) ) {
		state.endDate = toMoment( state.startDate ).add( HALF_HOUR_IN_SECONDS, 'seconds' );
	}

	return {
		...prevState,
		...adjustEventPropertiesFromMoment( state ),
	};
}

function adjustEventPropertiesFromMoment( state ) {
	return {
		...state,
		multiDay: ! isSameDay( state.endDate, state.startDate ),
		allDay: isAllDay( state.startDate, state.endDate ),
		startDate: toDateTime( state.startDate ),
		endDate: toDateTime( state.endDate ),
	};
}

function getAllDayDates( prevState ) {
	const { startDate } = prevState;
	const start = setTimeInSeconds( toMoment( startDate ), 0 );
	const end = toMoment( startDate ).endOf( 'day' );
	const state = {
		startDate: toDateTime( start ),
		allDay: true,
		multiDay: false,
		endDate: toDateTime( end ),
	};

	return {
		...prevState,
		...state,
	};
}

function getMultiDayDates( prevState ) {
	const { startDate, endDate } = prevState;
	const start = toMoment( startDate );
	const state = {
		...prevState,
		multiDay: true,
		allDay: false,
		startDate,
		endDate,
	};

	if ( isSameDay( start, toMoment( state.endDate ) ) ) {
		// Add 3 days if we are going to set a dynamic range when start and end are on the same day
		const range = DAY_IN_SECONDS * 3;
		state.endDate = toDateTime(
			toMoment( startDate ).add( range, 'seconds' ),
		);
	}

	return state;
}

export function setOrganizers( prevState, organizer ) {
	return {
		...prevState,
		eventOrganizers: organizer,
	}
}

export function addOrganizers( prevState, organizer ) {
	const { eventOrganizers } = prevState;
	return {
		...prevState,
		eventOrganizers: [ ...eventOrganizers, organizer ],
	};
}

export function setCurrencySymbol( prevState, symbol ) {
	return {
		...prevState,
		eventCurrencySymbol: symbol,
	};
}

export function setCurrencyPosition( prevState, hasPosition ) {
	const position = isTruthy( hasPosition ) ? 'suffix' : 'prefix';
	return {
		...prevState,
		setCurrencyPosition: position,
	};
}

export function setWebsiteUrl( prevState, url ) {
	return {
		...prevState,
		eventUrl: url,
	};
}