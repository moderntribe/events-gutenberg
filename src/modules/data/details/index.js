/**
 * External Dependencies
 */
import moment from 'moment/moment';
import { __ } from '@wordpress/i18n';
import { isObject } from 'lodash';

/**
 * Wordpress dependencies
 */
const { data } = wp;
const { registerStore } = data;

/**
 * Internal dependencies
 */
import { roundTime, toDateTime } from 'utils/moment';
import { HALF_HOUR_IN_SECONDS } from 'utils/time';

import * as reducers from './reducers';
import { getSetting } from 'editor/settings';
import { isTruthy } from 'utils/string';

export const DEFAULT_STATE = {
	multiDay: false,
	allDay: false,
	startDate: toDateTime( roundTime( moment() ) ),
	endDate: toDateTime( roundTime( moment() ).add( HALF_HOUR_IN_SECONDS, 'seconds' ) ),
	timezone: 'UTC',
	dateTimeRangeSeparator: getSetting( 'dateTimeSeparator', __( ' @ ', 'events-gutenberg' ) ),
	timeRangeSeparator: getSetting( 'timeRangeSeparator', __( ' - ', 'events-gutenberg' ) ),
	currencyPosition: isTruthy( getSetting( 'reverseCurrencyPosition', 0 ) ) ? 'suffix' : 'prefix',
	eventCurrencySymbol: getSetting( 'defaultCurrencySymbol', __( '$', 'events-gutenberg' ) ),
	dashboardOpen: false,
	eventOrganizers: [],
	url: undefined,
};

export const STORE_NAME = 'tec.details';

const details = {
	reducer( state = DEFAULT_STATE, action ) {
		switch ( action.type ) {
			case 'SET_INITIAL_STATE': {
				return reducers.setInitialState( state, action.values );
			}

			case 'SET_MULTI_DAY': {
				return reducers.setMultiDay( state, action.multiDay );
			}

			case 'SET_ALL_DAY': {
				return reducers.setAllDay( state, action.allDay );
			}

			case 'SET_TIME_ZONE': {
				return {
					...state,
					timezone: action.timezone,
				};
			}

			case 'SET_START_DATE': {
				return reducers.setOnlyDateForStart( state, action.date );
			}

			case 'SET_END_DATE': {
				return reducers.setOnlyDateForEnd( state, action.date );
			}

			case 'SET_START_TIME': {
				return reducers.setStartTime( state, action.seconds );
			}

			case 'SET_END_TIME': {
				return reducers.setEndTime( state, action.seconds );
			}

			case 'SET_ORGANIZERS': {
				return reducers.setOrganizers( state, action.organizers );
			}

			case 'ADD_ORGANIZER': {
				return reducers.addOrganizers( state, action.organizer );
			}

			case 'REMOVE_ORGANIZER': {
				return reducers.removeOrganizer( state, action.organizer );
			}

			case 'MAYBE_REMOVE_ORGANIZER': {
				return reducers.maybeRemoveOrganizer( state, action.organizer );
			}

			case 'REPLACE_ORGANIZERS': {
				return reducers.replaceOrganizers( state, action.organizers );
			}

			case 'SET_CURRENCY_SYMBOL': {
				return reducers.setCurrencySymbol( state, action.symbol );
			}

			case 'SET_CURRENCY_POSITION': {
				return reducers.setCurrencyPosition( state, action.position );
			}

			case 'SET_WEBSITE_URL': {
				return reducers.setWebsiteUrl( state, action.url );
			}

			case 'SET_DATE_TIME_SEPARATOR': {
				return {
					...state,
					dateTimeRangeSeparator: action.separator,
				};
			}

			case 'SET_TIME_RANGE_SEPARATOR': {
				return {
					...state,
					timeRangeSeparator: action.separator,
				};
			}

			case 'TOGGLE_DASHBOARD': {
				return {
					...state,
					dashboardOpen: ! state.dashboardOpen,
				};
			}

			case 'CLOSE_DASHBOARD': {
				return {
					...state,
					dashboardOpen: false,
				};
			}

			default: {
				return state;
			}
		}
	},
	actions: {
		addOrganizer( organizer ) {
			return {
				type: 'ADD_ORGANIZER',
				organizer,
			};
		},
		removeOrganizer( organizer ) {
			return {
				type: 'REMOVE_ORGANIZER',
				organizer,
			};
		},
		maybeRemoveOrganizer( organizer ) {
			return {
				type: 'MAYBE_REMOVE_ORGANIZER',
				organizer,
			}
		},
		replaceOrganizers( organizers ) {
			return {
				type: 'REPLACE_ORGANIZERS',
				organizers,
			};
		},
	},
	selectors: {
		getOrganizers( state ) {
			const { eventOrganizers } = state;
			return eventOrganizers
				.filter( isObject )
				.map( ( item ) => item.id );
		},
	},
};

export const store = registerStore( STORE_NAME, details );
