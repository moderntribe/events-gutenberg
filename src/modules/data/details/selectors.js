/**
 * External dependencies
 */
import { createSelector } from 'reselect';
import { DEFAULT_STATE } from './reducers/details';

const blockSelector = ( state, props ) => state.details[ props.name ];

export const getPostType = createSelector(
	[ blockSelector ],
	( block ) => block ? block.type : DEFAULT_STATE.type
);

export const getLoading = createSelector(
	[ blockSelector ],
	( block ) => block ? block.loading : DEFAULT_STATE.loading,
);

export const getDetails = createSelector(
	[ blockSelector ],
	( block ) => block ? block.details : DEFAULT_STATE.details,
);
