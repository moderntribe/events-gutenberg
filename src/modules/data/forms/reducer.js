/**
 * Internal dependencies
 */
import * as types from './types';
import { form } from './reducers';

export default ( state = {}, action ) => {
	switch ( action.type ) {
		case types.ADD_FORM:
		case types.CLEAR_FORM:
		case types.SET_FORM_FIELDS:
		case types.CREATE_FORM_DRAFT:
		case types.EDIT_FORM_ENTRY:
		case types.SUBMIT_FORM:
			return {
				...state,
				[ action.payload.id ]: form( state[ action.payload.id ], action ),
			};
		default:
			return state;
	}
};
