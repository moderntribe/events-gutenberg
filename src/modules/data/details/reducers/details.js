/**
 * Internal dependencies
 */
import { EVENT } from 'editor/post-types';
import { types } from 'data/details';

export const DEFAULT_STATE = {
	loading: false,
	details: {},
	postType: EVENT,
};

export default ( state = DEFAULT_STATE, action ) => {
	switch ( action.type ) {
		case types.SET_DETAILS:
			return {
				...state,
				details: action.payload.details,
			};
		case types.SET_DETAILS_POST_TYPE:
			return {
				...state,
				postType: action.payload.postType,
			};
		case types.SET_DETAILS_LOADING:
			return {
				...state,
				loading: action.payload.loading,
			};
		default:
			return state;
	}
};
