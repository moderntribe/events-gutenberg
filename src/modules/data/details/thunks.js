/**
 * External dependencies
 */
import { isEmpty } from 'lodash';

/**
 * Internal dependencies
 */
import * as actions from './actions';
import * as selectors from './selectors';
import { actions as requestActions } from 'data/request';

export const fetchDetails = ( id ) => ( dispatch, getState ) => {
	const state = getState();
	const props = { name: id };
	const loading = selectors.getLoading( state, props );
	const details = selectors.getDetails( state, props );

	if ( ! isEmpty( details ) || loading ) {
		return;
	}

	const postType = selectors.getPostType( state, props );
	const options = {
		path: `${ postType }/${ id }`,
		actions: {
			start: () => dispatch( actions.enableLoading( id ) ),
			success: ( { body } ) => {
				dispatch( actions.setDetails( id, body ) );
				dispatch( actions.disableLoading( id ) );
			},
			error: () => dispatch( actions.disableLoading( id ) ),
		},
	};

	dispatch( requestActions.wpRequest( options ) );
};
