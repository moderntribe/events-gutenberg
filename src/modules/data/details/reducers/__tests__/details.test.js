/**
 * Internal dependencies
 */
import { actions } from 'data/details';
import reducer, { DEFAULT_STATE } from 'data/details/reducers/details';

describe( '[STORE] - details reducer', () => {
	it( 'Should return default state', () => {
		expect( reducer( undefined, {} ) ).toEqual( DEFAULT_STATE );
	} );

	it( 'Should set the details', () => {
		expect( reducer( DEFAULT_STATE, actions.setDetails( 20, { title: 'Modern Tribe' } ) ) )
			.toMatchSnapshot();
	} );

	it( 'Should enable the loading', () => {
		expect( reducer( DEFAULT_STATE, actions.enableIsLoading( 20 ) ) )
			.toMatchSnapshot();
	} );

	it( 'Should set the post type', () => {
		expect( reducer( DEFAULT_STATE, actions.setPostType( 20, 'tribe_organizers' ) ) )
			.toMatchSnapshot();
	} );
} );
