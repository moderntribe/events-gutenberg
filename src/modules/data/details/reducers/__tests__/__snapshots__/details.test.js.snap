// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[STORE] - details reducer Should enable the loading 1`] = `
Object {
  "details": Object {},
  "loading": true,
  "postType": "tribe_events",
}
`;

exports[`[STORE] - details reducer Should set the details 1`] = `
Object {
  "details": Object {
    "title": "Modern Tribe",
  },
  "loading": false,
  "postType": "tribe_events",
}
`;

exports[`[STORE] - details reducer Should set the post type 1`] = `
Object {
  "details": Object {},
  "loading": false,
  "postType": "tribe_organizers",
}
`;
